Add blogpost on understanding breakage (#26)
haskell/blog.haskell.org: content/investigating-lang-ext-semantics/index.md

+++
title = "Analyzing language extension semantics"
date = 2025-11-26
[taxonomies]
authors = ["Jappie Klooster"]
categories = ["Haskell Foundation"]
tags = ["Community", "Stability"]
+++

> We analyzed the head.hackage patches to understand 

Reviewer: What is `head.hackage`? It's not defined to the reader at this point of the document.

why code breaks on new GHC releases.
Surprisingly, most breakage wasn’t caused by 
Template Haskell—it came from deeper semantic changes in language extensions.
This post walks through the main categories of breakage, 
why they happened, and what they tell us about long-term stability.
If you care about a smoother upgrade path for Haskell users, 
we invite you to participate in the Stability Working Group.

The [Haskell Foundation Stability Working Group](https://blog.haskell.org/stability-working-group/) 
is interested in understanding
why breakage occurs.
This is an extension of our initial [investigation](https://jappie.me/analyzing-haskell-stability.html).
We've recently done further analysis on [head.hackage](https://ghc.gitlab.haskell.org/head.hackage/),
and learned surprisingly enough that the root cause
of a lot of breakage isn't Template Haskell,
but seems to be from language extension semantics[^meaning].
We're doing this investigation to better understand where the Haskell

> Foundation stability working group should focus it's efforts.

Reviewer: ```suggestion
Foundation stability working group should focus its efforts.
```

This gave us the following table:

| Name                                | Cause                                        | Had warnings? |
|-------------------------------------+----------------------------------------------+---------------|
| Cabal-2.4.1.0.patch                 | simplified subsumption                       | no            |
| Cabal-3.0.2.0.patch                 | simplified subsumption                       | no            |
| Cabal-3.2.1.0.patch                 | simplified subsumption                       | no            |
| data-r-tree-0.6.0.patch             | parser change (see 1)                        | no            |
| drinkery-0.4.patch                  | simplified subsumption                       | no            |
| ghc-lib-parser-9.8.1.20231121.patch | rename forall identifiers (2)                | yes           |
| hgeometry-ipe-0.13.patch            | Instances moved due to splice enforcement    | no            |
| singletons-3.0.2.patch              | add TypeAbstractions as a language extension | yes           |
| singletons-base-3.1.1.patch         | add TypeAbstractions as a language extension | yes           |

> | vector-space-0.16.patch             | * is type (4)                                | yes           |

Reviewer: This lone `*` makes me frown. There is a possibility it is throwing the whole Markdown renderer off.
<img width="1292" height="291" alt="image" src="https://github.com/user-attachments/assets/3af11654-0bd7-4db0-bc94-2eeb490bb0ff" />

@jappeace It should probably be escaped?

Reviewer: @jappeace welp it looks like that wasn't it. :/ Do you have a Markdown renderer on your computer that you could check against?

Author: no I asked the chatbot, it was +

Author: https://github.com/haskell/blog.haskell.org/pull/26/commits/e16b382c7174502ed0c6f13517522d617bd4d35b

`th-compat-0.1.4.patch` was miscounted so I left that out.
Simplified subsumption appears a lot but 3 are for Cabal,
so it's only 2 real occurrences.
We expect that to appear a lot however,
because it was one of *the* motivating changes for a [stability working group](https://blog.haskell.org/stability-working-group/).

## Simplified subsumption
For the blissfully ignorant reader simplified subsumption causes you
to do this under certain existential conditions:

> ```haskell

Reviewer: If you want to highlight the changes instead of the syntax, you should use `diff` instead of `haskell`. This will help the readers understand what has actually changed.

--- a/Distribution/Simple/Utils.hs
+++ b/Distribution/Simple/Utils.hs
@@ -1338,7 +1338,7 @@ withTempFileEx opts tmpDir template action =
     (\(name, handle) -> do hClose handle
                            unless (optKeepTempFiles opts) $
                              handleDoesNotExist () . removeFile $ name)
-    (withLexicalCallStack (uncurry action))
+    (withLexicalCallStack (\x -> uncurry action x))
 
```

> You've to insert a lambda, which apparently signifies some performance impact.

Reviewer: ```suggestion
You've to insert a lambda, which apparently has some performance impact.
```

This went wild with [Yesod stacks](https://www.yesodweb.com/book), 
whose code generation helpfully created 
the database alias in the template:
```haskell
type DB a = forall (m :: Type -> Type).
    (MonadUnliftIO m) => ReaderT SqlBackend m a
```

So anything that now uses a query has to insert those lambdas,
as you can imagine this would be in quite a few places for non-trivial commercial code bases.
Which caused many issues for commercial users.
You can just delete those aliases to solve the problem.
Alternatively you can just enable the language extension: [DeepSubsumption](https://downloads.haskell.org/~ghc/9.12.2/docs/users_guide/exts/rank_polymorphism.html#extension-DeepSubsumption).
Which restores the original behavior.

## Moving of instances due to Template Haskell
This change forces you to put the instances above the splice where

> it's being used in the same module.

Reviewer: ```suggestion
they are being used in the same module.
```

A dear colleague decided to generate instances in Template Haskell.
That was quite the puzzle!
I asked the GHC devs why they did this,
and it turns out this was a soundness issue in the typechecker.
Here, soundness means the type system can't be tricked into allowing invalid programs.
So the community is better off, despite this causing a fair bit of work.

> ```haskell

Reviewer: Similarly, `diff` should be used.

But semantics did change! 
Actually I don't think we expected to find something like this at all.
I'm not sure how relevant this is to discuss further because it appears
quite rare for someone to do this.
You can enable [StrictData](https://downloads.haskell.org/ghc/latest/docs/users_guide/exts/strict.html#extension-StrictData)
in your Cabal file and delete all those bangs!

## (2) Rename forall identifiers
This changes the forall identifier into a keyword at term level.
It already was at the type level.
The issue is discussed [here](https://gitlab.haskell.org/ghc/ghc/-/issues/23719)

```haskell
 hintExplicitForall :: Located Token -> P ()
 hintExplicitForall tok = do
-    forall   <- getBit ExplicitForallBit
+    forAll   <- getBit ExplicitForallBit
     rulePrag <- getBit InRulePragBit
-    unless (forall || rulePrag) $ addError $ mkPlainErrorMsgEnvelope (getLoc tok) $
+    unless (forAll || rulePrag) $ addError $ mkPlainErrorMsgEnvelope (getLoc tok) $
       (PsErrExplicitForall (isUnicode tok))
```

## TypeAbstractions

From what I understand from the [manual](https://ghc.gitlab.haskell.org/ghc/doc/users_guide/exts/type_abstractions.html#type-abstractions)

> Is that part of the syntax for type abstractions landed in GHC 9.2,

Reviewer: ```suggestion
is that part of the syntax for type abstractions landed in GHC 9.2,
```

however 9.8 and onwards requires you to enable this language extension.

> This appears to because certain new functionality was introduced behind an

Reviewer: ```suggestion
This appeared because certain new functionality was introduced behind an
```

> old language extension flag, according to [this proposal](https://github.com/ghc-proposals/ghc-proposals/blob/master/proposals/0448-type-variable-scoping.rst#4type-arguments-in-constructor-patterns)

Reviewer: ```suggestion
old language extension flag, according to [this proposal](https://github.com/ghc-proposals/ghc-proposals/blob/master/proposals/0448-type-variable-scoping.rst#4type-arguments-in-constructor-patterns).
```

|-------------------------------------|----------------------------------------------|---------------|
| Cabal-2.4.1.0.patch                 | simplified subsumption                       | no            |
| Cabal-3.0.2.0.patch                 | simplified subsumption                       | no            |
| Cabal-3.2.1.0.patch                 | simplified subsumption                       | no            |
| data-r-tree-0.6.0.patch             | parser change (see 1)                        | no            |
| drinkery-0.4.patch                  | simplified subsumption                       | no            |
| ghc-lib-parser-9.8.1.20231121.patch | rename forall identifiers (2)                | yes           |
| hgeometry-ipe-0.13.patch            | Instances moved due to splice enforcement    | no            |
| singletons-3.0.2.patch              | add TypeAbstractions as a language extension | yes           |
| singletons-base-3.1.1.patch         | add TypeAbstractions as a language extension | yes           |
| vector-space-0.16.patch             | Star is type (4)                             | yes           |

|--

> 

Reviewer: ```suggestion
```

disallowing `!` before parens.
Here the bang `!` indicates strict fields.
Technically this doesn't fit into the category
because the core language isn't a language extension.
But semantics did change! 
Actually I don't think we expected to find something like this at all.
I'm not sure how relevant this is to discuss further because it appears
quite rare for someone to do this.
You can enable [StrictData](https://downloads.haskell.org/ghc/latest/docs/users_guide/exts/strict.html#extension-StrictData)
in your Cabal file and delete all those bangs!

## (2) Rename forall identifiers
This changes the forall identifier into a keyword at term level.
It already was at the type level.
The issue is discussed [here](https://gitlab.haskell.org/ghc/ghc/-/issues/23719)

```diff
 hintExplicitForall :: Located Token -> P ()
 hintExplicitForall tok = do
-    forall   <- getBit ExplicitForallBit
+    forAll   <- getBit ExplicitForallBit
     rulePrag <- getBit InRulePragBit
-    unless (forall || rulePrag) $ addError $ mkPlainErrorMsgEnvelope (getLoc tok) $
+    unless (forAll || rulePrag) $ addError $ mkPlainErrorMsgEnvelope (getLoc tok) $
       (PsErrExplicitForall (isUnicode tok))
```

> ## TypeAbstractions

Reviewer: ```suggestion
## (3) TypeAbstractions
```

From what I understand from the [manual](https://ghc.gitlab.haskell.org/ghc/doc/users_guide/exts/type_abstractions.html#type-abstractions)
is that part of the syntax for type abstractions landed in GHC 9.2,
however 9.8 and onwards requires you to enable this language extension.
This appeared because certain new functionality was introduced behind an
old language extension flag, according to [this proposal](https://github.com/ghc-proposals/ghc-proposals/blob/master/proposals/0448-type-variable-scoping.rst#4type-arguments-in-constructor-patterns). It says we don't want to introduce new functionality behind established extensions,
so that's why we require TypeAbstractions now, 
where previously ScopedTypeVariables and TypeApplications were enough.

This extension enables you to bind type variables in pattern matches.
I don't know why this happened like this, but it happened in 2023:

```diff
+-- Invisible type binders in type declarations, such as
+--
+--   type family Sing @k
+--
+-- require the TypeAbstractions extension.
+#if __GLASGOW_HASKELL__ >= 909
+{-# LANGUAGE TypeAbstractions #-}
+#endif
+
```

> ## Star is type

Reviewer: ```suggestion
## (4) Star is type
```

This change was announced via a warning. 
It tells users to write `Type` instead of `*` for kinds representing types.
A kind is essentially the type of a type,
and as a concept is used for type-level programming type safety.

```diff
-  type Basis v :: *
+  type Basis v :: Type
```

> 

Reviewer: ```suggestion

```

## Conclusion

Often these breakages are annoying and frustrating.
But if we look deeper, we find that each of them has
a little story
and good reasons for being introduced.
If you find this all as interesting as I do,
please consider joining some of the stability
working group meetings!

> 

Reviewer: ```suggestion
```
